widgets/url-sync/index.js

```diff
@@ -0,0 +1,172 @@
+
+var isEqual = require('lodash/lang/isEqual');
+var merge = require('lodash/object/merge');
+var forEach = require('lodash/collection/forEach');
+
+var getNow = Date.now ? function() { return Date.now(); } : function() { return (new Date()).getTime(); };
+
+var timerMaker = function(t0){
+  var t = t0;
+  return function timer(){
+    var now = getNow();
+    var delta = now - t;
+    t = now;
+    return delta;
+  };
+};
+
+/**
+ * @typedef {object} UrlUtil
+ * @property {string} character the character used in the url
+ * @property {function} onpopstate add an event listener for the URL change
+ * @property {function} pushState creates a new entry in the browser history
+ * @property {function} replaceState update the current entry of the browser history
+ * @property {function} readUrl reads the query string of the parameters
+ */
+
+/**
+ * Handles the legacy browsers
+ * @type {UrlUtil}
+ */
+var hashUrlUtils = {
+  character: '#',
+  onpopstate: function(cb){
+    if (window.addEventListener) {
+      window.addEventListener('hashchange', cb);
+    } else if (window.attachEvent) {
+      window.attachEvent('onhashchange', cb); 
+    }
+  },
+  pushState: function(qs){
+    window.location.assign(this.character + qs);
+  },
+  replaceState: function(qs){
+    window.location.replace(this.character + qs);
+  },
+  readUrl: function(){
+    return window.location.hash.slice(1);
+  }
+};
+
+/**
+ * Handles the modern API
+ * @type {UrlUtil}
+ */
+var modernUrlUtils = {
+  character: '?',
+  onpopstate: function(cb){
+    window.addEventListener('popstate', cb);
+  },
+  pushState: function(qs){
+    window.history.pushState(null, '', this.character + qs);
+  },
+  replaceState: function(qs){
+    window.history.replaceState(null, '', this.character + qs);
+  },
+  readUrl: function(){
+    return window.location.search.slice(1);
+  }
+};
+
+/**
+ * Instanciate a url sync widget. This widget let you synchronize the search
+ * parameters with the URL. It can operate with legacy API and hash or it can use
+ * the modern history API. By default, it will use the modern API, but if you are
+ * looking for compatibility with IE8 and IE9, then you should set 'useHash' to
+ * true.
+ * @class
+ * @param {object} urlUtils an object containing the function to read, watch the changes
+ * and update the URL.
+ * @param {number} threshold time in ms after which a new state is created in the browser
+ * history. The default value is 700.
+ * @param {string[]} trackedParameters parameters that will be synchronized in the
+ * URL. By default, it will track the query, all the refinable attribute (facets and numeric
+ * filters), the index and the page.
+ * @param {boolean} useHash if set to true, the url will be hash based. Otherwise,
+ * it'll use the query parameters using the modern history API.
+ */
+class URLSync {
+  constructor(urlUtils, options) {
+    this.__initLast = true;
+    this.urlUtils = urlUtils;
```

Comment:
I am interested because I do not know anything about Symbol (or the use case).

The only thing I would like to do right now is hide or _prefix properties that will be passed to the api user. And right now there's no way for him to understand that it's somehow "private"